### PR TITLE
Initial Docker rewrite

### PR DESCRIPTION
**Pull request by sergystepanov**
Related to #164.

### CODE REVIEW

**Comment by sergystepanov** [comment]
Doesn't work on Windows.
Needs RTFM.

**Comment by sergystepanov** [comment]
@giongto35, could you describe a bit your cloudretro.io Docker config differences from the Makefile? If any. You just run it in host mode?

**Comment by giongto35** [comment]
Yes, it's better with host mode. I intended to make that change as well.
So you are going to rewrite it with Docker Compose with 2 separate containers coord + worker in host mode right? (y) Thank you so much for it.

**Comment by sergystepanov** [comment]
It's not working in any mode for me (: Wanted to try to fix it and close the ticket with optimized Docker image.
So, there is no special configs. I see.
Docker compose is a canonical way of that stuff in the Makefile and an example how you can run it more easily without Docker run and k8s insanity (e.g. `docker-compose scale coordinator=1 worker=10`).
Not sure about whether to use separate containers or one for both. How do you see it?

Maybe leave k8s for "production use" with scaling, replicas and such stuff, and compose for some local hacking?

**Comment by sergystepanov** [comment]
And, maybe, to add gh Docker release images for the app. Need reading a manual. (:

**Comment by giongto35** [comment]
Im debugging the docker issue. Seems like it is at server ping. At first, the browser can curl worker localhost:9000/echo but the next call after reloading fail. Can you check if it is this call succeed or fail on browser?
"http://localhost:9000/echo: 23" 
or "http://localhost:9000/echo: 10000" (timeout)
----
Regarding k8s, I want to setup K8S even on local last time but I failed to expose port range on service (for setting up peer connection)
https://github.com/kubernetes/kubernetes/issues/23864 . Seems like other people have same issue with that.

**Comment by sergystepanov** [comment]
image
Made the image size 132 Mb so far with a normal distro.

Echo is seems to be working
image
image

**Comment by sergystepanov** [comment]
Dockerized it doesn't work on Windows with Linux containers, with WSL2 backend also.
Native Ubuntu can run it out of a box:
Screenshot from 2020-06-06 11-57-12

**Comment by sergystepanov** [comment]
I'll try to fix ICE problems and leave all new Docker issues for someone who runs and devs on Linux natively.

**Comment by sergystepanov** [comment]
Own notes.
Ub:
```
sergu@sergu-laptop:~/cloud-game$ sudo docker-compose up
Creating network "cloud-game_default" with the default driver
Building cloud-game
Step 1/12 : FROM golang:1.14 AS build
1.14: Pulling from library/golang
376057ac6fa1: Pull complete
5a63a0a859d8: Pull complete
496548a8c952: Pull complete
2adae3950d4d: Pull complete
039b991354af: Pull complete
6b823afb12d9: Pull complete
30b9d62bd869: Pull complete
Digest: sha256:4d58164df51901679307e547ffed19837e63a1d1b9ed80453b8d88749b6033f4
Status: Downloaded newer image for golang:1.14
 ---> 05feda542433
Step 2/12 : WORKDIR /go/src/github.com/giongto35/cloud-game/
 ---> Running in 8fb50f066611
Removing intermediate container 8fb50f066611
 ---> 25ae0f377b8e
Step 3/12 : RUN apt-get update &&     apt-get install pkg-config libvpx-dev libopus-dev libopusfile-dev -y
 ---> Running in 9cfc446f3811
Get:1 http://security.debian.org/debian-security buster/updates InRelease [65.4 kB]
Get:2 http://deb.debian.org/debian buster InRelease [121 kB]
Get:3 http://deb.debian.org/debian buster-updates InRelease [49.3 kB]
Get:4 http://security.debian.org/debian-security buster/updates/main amd64 Packages [201 kB]
Get:5 http://deb.debian.org/debian buster/main amd64 Packages [7905 kB]
Get:6 http://deb.debian.org/debian buster-updates/main amd64 Packages [7376 B]
Fetched 8350 kB in 4s (2103 kB/s)
Reading package lists...
Reading package lists...
Building dependency tree...
Reading state information...
pkg-config is already the newest version (0.29-6).
The following additional packages will be installed:
  libogg-dev libogg0 libopus0 libopusfile0 libssl-dev libvpx5
Suggested packages:
  opus-tools libssl-doc
The following NEW packages will be installed:
  libogg-dev libogg0 libopus-dev libopus0 libopusfile-dev libopusfile0
  libssl-dev libvpx-dev libvpx5
0 upgraded, 9 newly installed, 0 to remove and 0 not upgraded.
Need to get 4286 kB of archives.
After this operation, 15.9 MB of additional disk space will be used.
Get:1 http://deb.debian.org/debian buster/main amd64 libogg0 amd64 1.3.2-1+b1 [21.3 kB]
Get:2 http://deb.debian.org/debian buster/main amd64 libogg-dev amd64 1.3.2-1+b1 [200 kB]
Get:3 http://deb.debian.org/debian buster/main amd64 libopus0 amd64 1.3-1 [191 kB]
Get:4 http://deb.debian.org/debian buster/main amd64 libopus-dev amd64 1.3-1 [238 kB]
Get:5 http://deb.debian.org/debian buster/main amd64 libopusfile0 amd64 0.9+20170913-1 [38.5 kB]
Get:6 http://deb.debian.org/debian buster/main amd64 libopusfile-dev amd64 0.9+20170913-1 [58.9 kB]
Get:7 http://deb.debian.org/debian buster/main amd64 libssl-dev amd64 1.1.1d-0+deb10u3 [1793 kB]
Get:8 http://deb.debian.org/debian buster/main amd64 libvpx5 amd64 1.7.0-3+deb10u1 [800 kB]
Get:9 http://deb.debian.org/debian buster/main amd64 libvpx-dev amd64 1.7.0-3+deb10u1 [944 kB]
debconf: delaying package configuration, since apt-utils is not installed
Fetched 4286 kB in 2s (2607 kB/s)
Selecting previously unselected package libogg0:amd64.
(Reading database ... 15412 files and directories currently installed.)
Preparing to unpack .../0-libogg0_1.3.2-1+b1_amd64.deb ...
Unpacking libogg0:amd64 (1.3.2-1+b1) ...
Selecting previously unselected package libogg-dev:amd64.
Preparing to unpack .../1-libogg-dev_1.3.2-1+b1_amd64.deb ...
Unpacking libogg-dev:amd64 (1.3.2-1+b1) ...
Selecting previously unselected package libopus0:amd64.
Preparing to unpack .../2-libopus0_1.3-1_amd64.deb ...
Unpacking libopus0:amd64 (1.3-1) ...
Selecting previously unselected package libopus-dev:amd64.
Preparing to unpack .../3-libopus-dev_1.3-1_amd64.deb ...
Unpacking libopus-dev:amd64 (1.3-1) ...
Selecting previously unselected package libopusfile0.
Preparing to unpack .../4-libopusfile0_0.9+20170913-1_amd64.deb ...
Unpacking libopusfile0 (0.9+20170913-1) ...
Selecting previously unselected package libopusfile-dev.
Preparing to unpack .../5-libopusfile-dev_0.9+20170913-1_amd64.deb ...
Unpacking libopusfile-dev (0.9+20170913-1) ...
Selecting previously unselected package libssl-dev:amd64.
Preparing to unpack .../6-libssl-dev_1.1.1d-0+deb10u3_amd64.deb ...
Unpacking libssl-dev:amd64 (1.1.1d-0+deb10u3) ...
Selecting previously unselected package libvpx5:amd64.
Preparing to unpack .../7-libvpx5_1.7.0-3+deb10u1_amd64.deb ...
Unpacking libvpx5:amd64 (1.7.0-3+deb10u1) ...
Selecting previously unselected package libvpx-dev:amd64.
Preparing to unpack .../8-libvpx-dev_1.7.0-3+deb10u1_amd64.deb ...
Unpacking libvpx-dev:amd64 (1.7.0-3+deb10u1) ...
Setting up libogg0:amd64 (1.3.2-1+b1) ...
Setting up libogg-dev:amd64 (1.3.2-1+b1) ...
Setting up libopus0:amd64 (1.3-1) ...
Setting up libssl-dev:amd64 (1.1.1d-0+deb10u3) ...
Setting up libopusfile0 (0.9+20170913-1) ...
Setting up libvpx5:amd64 (1.7.0-3+deb10u1) ...
Setting up libopus-dev:amd64 (1.3-1) ...
Setting up libvpx-dev:amd64 (1.7.0-3+deb10u1) ...
Setting up libopusfile-dev (0.9+20170913-1) ...
Processing triggers for libc-bin (2.28-10) ...
Removing intermediate container 9cfc446f3811
 ---> af74d247774f
Step 4/12 : COPY go.mod go.sum ./
 ---> bf21d25cedee
Step 5/12 : RUN go mod download
 ---> Running in 14edad64696d
Removing intermediate container 14edad64696d
 ---> 404cf139a40c
Step 6/12 : COPY ./ ./
 ---> 50f46cd0aab0
Step 7/12 : RUN go install ./cmd/coordinator &&     go install ./cmd/worker
 ---> Running in 7e46b9a38a58
Removing intermediate container 7e46b9a38a58
 ---> b4af61402341
Step 8/12 : FROM debian:10-slim
10-slim: Pulling from library/debian
afb6ec6fdc1c: Pull complete
Digest: sha256:47bde1681ab5258dceb05ffcd7d8ef1d367ce5c6631075599fb74ab2101cd151
Status: Downloaded newer image for debian:10-slim
 ---> 108d75da320f
Step 9/12 : RUN apt-get update &&     apt-get install libvpx5 libopus0 libopusfile0 -y &&     rm -rf /var/lib/apt/lists/*
 ---> Running in 75e8f5d7873b
Get:1 http://deb.debian.org/debian buster InRelease [121 kB]
Get:2 http://security.debian.org/debian-security buster/updates InRelease [65.4 kB]
Get:3 http://deb.debian.org/debian buster-updates InRelease [49.3 kB]
Get:4 http://security.debian.org/debian-security buster/updates/main amd64 Packages [201 kB]
Get:5 http://deb.debian.org/debian buster/main amd64 Packages [7905 kB]
Get:6 http://deb.debian.org/debian buster-updates/main amd64 Packages [7376 B]
Fetched 8350 kB in 4s (1910 kB/s)
Reading package lists...
Reading package lists...
Building dependency tree...
Reading state information...
The following additional packages will be installed:
  libogg0 libssl1.1
Suggested packages:
  opus-tools
The following NEW packages will be installed:
  libogg0 libopus0 libopusfile0 libssl1.1 libvpx5
0 upgraded, 5 newly installed, 0 to remove and 0 not upgraded.
Need to get 2590 kB of archives.
After this operation, 7023 kB of additional disk space will be used.
Get:1 http://deb.debian.org/debian buster/main amd64 libogg0 amd64 1.3.2-1+b1 [21.3 kB]
Get:2 http://deb.debian.org/debian buster/main amd64 libopus0 amd64 1.3-1 [191 kB]
Get:3 http://deb.debian.org/debian buster/main amd64 libssl1.1 amd64 1.1.1d-0+deb10u3 [1538 kB]
Get:4 http://deb.debian.org/debian buster/main amd64 libopusfile0 amd64 0.9+20170913-1 [38.5 kB]
Get:5 http://deb.debian.org/debian buster/main amd64 libvpx5 amd64 1.7.0-3+deb10u1 [800 kB]
debconf: delaying package configuration, since apt-utils is not installed
Fetched 2590 kB in 2s (1700 kB/s)
Selecting previously unselected package libogg0:amd64.
(Reading database ... 6460 files and directories currently installed.)
Preparing to unpack .../libogg0_1.3.2-1+b1_amd64.deb ...
Unpacking libogg0:amd64 (1.3.2-1+b1) ...
Selecting previously unselected package libopus0:amd64.
Preparing to unpack .../libopus0_1.3-1_amd64.deb ...
Unpacking libopus0:amd64 (1.3-1) ...
Selecting previously unselected package libssl1.1:amd64.
Preparing to unpack .../libssl1.1_1.1.1d-0+deb10u3_amd64.deb ...
Unpacking libssl1.1:amd64 (1.1.1d-0+deb10u3) ...
Selecting previously unselected package libopusfile0.
Preparing to unpack .../libopusfile0_0.9+20170913-1_amd64.deb ...
Unpacking libopusfile0 (0.9+20170913-1) ...
Selecting previously unselected package libvpx5:amd64.
Preparing to unpack .../libvpx5_1.7.0-3+deb10u1_amd64.deb ...
Unpacking libvpx5:amd64 (1.7.0-3+deb10u1) ...
Setting up libogg0:amd64 (1.3.2-1+b1) ...
Setting up libssl1.1:amd64 (1.1.1d-0+deb10u3) ...
debconf: unable to initialize frontend: Dialog
debconf: (TERM is not set, so the dialog frontend is not usable.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (Can't locate Term/ReadLine.pm in @INC (you may need to install the Term::ReadLine module) (@INC contains: /etc/perl /usr/local/lib/x86_64-linux-gnu/perl/5.28.1 /usr/local/share/perl/5.28.1 /usr/lib/x86_64-linux-gnu/perl5/5.28 /usr/share/perl5 /usr/lib/x86_64-linux-gnu/perl/5.28 /usr/share/perl/5.28 /usr/local/lib/site_perl /usr/lib/x86_64-linux-gnu/perl-base) at /usr/share/perl5/Debconf/FrontEnd/Readline.pm line 7.)
debconf: falling back to frontend: Teletype
Setting up libopus0:amd64 (1.3-1) ...
Setting up libopusfile0 (0.9+20170913-1) ...
Setting up libvpx5:amd64 (1.7.0-3+deb10u1) ...
Processing triggers for libc-bin (2.28-10) ...
Removing intermediate container 75e8f5d7873b
 ---> 985b1e23c1e6
Step 10/12 : COPY --from=build /go/bin/ /
 ---> 570a7732f6f2
Step 11/12 : COPY ./web /web
 ---> 946ec455ffc8
Step 12/12 : EXPOSE 8000
 ---> Running in 5194853bcbf0
Removing intermediate container 5194853bcbf0
 ---> 015c291a4be7
Successfully built 015c291a4be7
Successfully tagged cloud-game:latest
WARNING: Image for service cloud-game was built because it did not already exist. To rebuild this image you must use `docker-compose build` or `docker-compose up --build`.
Creating cloud-game_cloud-game_1 ... done
Attaching to cloud-game_cloud-game_1
cloud-game_1  | I0606 11:51:35.516408       6 main.go:27] Initializing coordinator server
cloud-game_1  | I0606 11:51:35.516498       6 main.go:28] Coordinator configs {8800 http://localhost:8000   {6601 /coordinator false false}}
cloud-game_1  | I0606 11:51:35.516569       6 coordinator.go:103] Initializing Coordinator Server
cloud-game_1  | I0606 11:51:35.516579       6 coordinator.go:45] Starting monitoring server for coordinator
cloud-game_1  | I0606 11:51:35.516615       6 monitoring.go:81] Monitoring server is disabled via config
cloud-game_1  | I0606 11:51:35.521681       7 main.go:29] Initializing worker server
cloud-game_1  | I0606 11:51:35.521866       7 worker.go:47] Starting monitoring server for overwork
cloud-game_1  | I0606 11:51:35.521952       7 monitoring.go:53] Starting monitoring server at :6601
cloud-game_1  | I0606 11:51:35.521962       7 monitoring.go:75] Prometheus metric is enabled at :6601/worker/metrics
cloud-game_1  | I0606 11:51:35.527576       7 storage.go:25] Warn: Failed to create client: dialing: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.
cloud-game_1  | I0606 11:51:35.527611       7 worker.go:154] Listening at port: localhost: 9000
cloud-game_1  | I0606 11:51:35.527761       7 handlers.go:95] Worker connecting to coordinator: ws://localhost:8000/wso?zone=
cloud-game_1  | I0606 11:51:35.529520       6 handlers.go:87] Coordinator: A worker is connecting...
cloud-game_1  | I0606 11:51:35.529644       6 worker.go:41] Worker 6f9e1b72-d243-48ac-8cc0-b34f90498192] Generated worker ID
cloud-game_1  | I0606 11:51:35.529662       6 worker.go:41] Worker 6f9e1b72-d243-48ac-8cc0-b34f90498192] Address:127.0.0.1
cloud-game_1  | I0606 11:51:35.529678       6 worker.go:36] Worker 6f9e1b72-d243-48ac-8cc0-b34f90498192] Is public: false zone: 
cloud-game_1  | I0606 11:51:35.529680       7 handlers.go:72] Connected to coordinator successfully. &{0xc0003d49c0} <nil>
cloud-game_1  | I0606 11:51:35.529970       7 coordinator.go:45] Received serverID  6f9e1b72-d243-48ac-8cc0-b34f90498192
cloud-game_1  | I0606 11:51:41.312994       6 handlers.go:161] Coordinator: A user is connecting...
cloud-game_1  | I0606 11:51:41.313108       6 browser.go:35] Browser 4c5de699-5309-40f9-b52b-63e801bd3a23] Generated worker ID
cloud-game_1  | I0606 11:51:41.313150       6 browser.go:30] Browser 4c5de699-5309-40f9-b52b-63e801bd3a23] Get Room  Zone  From URL /ws?room_id=&zone=
cloud-game_1  | I0606 11:51:41.313184       6 browser.go:35] Browser 4c5de699-5309-40f9-b52b-63e801bd3a23] Send sync[http://localhost:9000/echo]http://localhost:9000/echo
cloud-game_1  | I0606 11:51:41.331346       7 worker.go:75] &{0xc00024a0a0 0xc000152100 {} 0x49b820 false false false false 0xc000074a40 {0xc000414000 map[] false false} map[Access-Control-Allow-Origin:[*]] true 0 -1 0 false false [] 0 [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] [0 0 0 0 0 0 0 0 0 0] [0 0 0] 0xc0000720e0 0} echo
cloud-game_1  | I0606 11:51:41.336753       6 browser.go:35] Browser 4c5de699-5309-40f9-b52b-63e801bd3a23] Latency mapmap[0xc00008e7e0:11]
cloud-game_1  | I0606 11:51:43.558467       6 browser.go:30] Browser 4c5de699-5309-40f9-b52b-63e801bd3a23] Received initwebrtc request -> relay to worker: 6f9e1b72-d243-48ac-8cc0-b34f90498192
cloud-game_1  | I0606 11:51:43.558843       7 coordinator.go:57] Received a request to createOffer from browser via coordinator
cloud-game_1  | I0606 11:51:43.558955       7 webrtc.go:115] === StartClient ===
cloud-game_1  | I0606 11:51:43.607992       7 webrtc.go:135] Add video track
cloud-game_1  | I0606 11:51:43.608606       7 webrtc.go:225] Created Offer
cloud-game_1  | I0606 11:51:43.608868       7 webrtc.go:186] OnIceCandidate: candidate:foundation 1 udp 2130706431 172.18.0.2 37974 typ host
cloud-game_1  | I0606 11:51:43.608967       7 coordinator.go:89] Start peerconnection 4c5de699-5309-40f9-b52b-63e801bd3a23
cloud-game_1  | I0606 11:51:43.609139       7 webrtc.go:186] OnIceCandidate: candidate:foundation 1 udp 1694498815 95.30.109.77 40393 typ srflx raddr 0.0.0.0 rport 40393
cloud-game_1  | I0606 11:51:43.609460       6 worker.go:41] Worker 6f9e1b72-d243-48ac-8cc0-b34f90498192] Received IceCandidate from worker -> relay to browser
cloud-game_1  | I0606 11:51:43.609628       6 worker.go:41] Worker 6f9e1b72-d243-48ac-8cc0-b34f90498192] Received IceCandidate from worker -> relay to browser
cloud-game_1  | I0606 11:51:43.609746       6 worker.go:41] Worker 6f9e1b72-d243-48ac-8cc0-b34f90498192] Received IceCandidate from worker -> relay to browser
cloud-game_1  | I0606 11:51:43.609856       6 browser.go:35] Browser 4c5de699-5309-40f9-b52b-63e801bd3a23] Received SDP from worker -> sending back to browser
cloud-game_1  | I0606 11:51:43.630474       6 browser.go:35] Browser 4c5de699-5309-40f9-b52b-63e801bd3a23] Received browser answered SDP -> relay to worker
cloud-game_1  | I0606 11:51:43.630755       7 coordinator.go:106] Received answer SDP from browser
cloud-game_1  | I0606 11:51:43.630976       7 webrtc.go:258] Set Remote Description
cloud-game_1  | I0606 11:51:43.631165       7 webrtc.go:170] ICE Connection State has changed: checking
cloud-game_1  | I0606 11:51:43.636289       6 browser.go:35] Browser 4c5de699-5309-40f9-b52b-63e801bd3a23] Received IceCandidate from browser -> relay to worker
cloud-game_1  | I0606 11:51:43.636562       7 coordinator.go:127] Received remote Ice Candidate from browser
cloud-game_1  | I0606 11:51:43.636609       7 webrtc.go:269] Decoded Ice: candidate:0 1 UDP 2122252543 192.168.1.121 39001 typ host
cloud-game_1  | I0606 11:51:43.636685       7 webrtc.go:277] Add Ice Candidate: candidate:0 1 UDP 2122252543 192.168.1.121 39001 typ host
cloud-game_1  | I0606 11:51:43.638317       6 browser.go:35] Browser 4c5de699-5309-40f9-b52b-63e801bd3a23] Received IceCandidate from browser -> relay to worker
cloud-game_1  | I0606 11:51:43.638398       6 browser.go:35] Browser 4c5de699-5309-40f9-b52b-63e801bd3a23] Received IceCandidate from browser -> relay to worker
cloud-game_1  | I0606 11:51:43.638447       6 browser.go:35] Browser 4c5de699-5309-40f9-b52b-63e801bd3a23] Received IceCandidate from browser -> relay to worker
cloud-game_1  | I0606 11:51:43.638668       7 coordinator.go:127] Received remote Ice Candidate from browser
cloud-game_1  | I0606 11:51:43.638701       7 webrtc.go:269] Decoded Ice: candidate:10 1 UDP 2122187007 172.18.0.1 55024 typ host
cloud-game_1  | I0606 11:51:43.638752       6 browser.go:35] Browser 4c5de699-5309-40f9-b52b-63e801bd3a23] Received IceCandidate from browser -> relay to worker
cloud-game_1  | I0606 11:51:43.638772       7 webrtc.go:277] Add Ice Candidate: candidate:10 1 UDP 2122187007 172.18.0.1 55024 typ host
cloud-game_1  | I0606 11:51:43.638866       7 coordinator.go:127] Received remote Ice Candidate from browser
cloud-game_1  | I0606 11:51:43.638902       7 webrtc.go:269] Decoded Ice: candidate:15 1 TCP 2105524479 192.168.1.121 9 typ host tcptype active
cloud-game_1  | I0606 11:51:43.639001       7 coordinator.go:127] Received remote Ice Candidate from browser
cloud-game_1  | I0606 11:51:43.639021       7 webrtc.go:269] Decoded Ice: candidate:16 1 TCP 2105393407 172.17.0.1 9 typ host tcptype active
cloud-game_1  | I0606 11:51:43.639082       7 coordinator.go:127] Received remote Ice Candidate from browser
cloud-game_1  | I0606 11:51:43.639102       7 webrtc.go:269] Decoded Ice: candidate:5 1 UDP 2122121471 172.17.0.1 44393 typ host
cloud-game_1  | I0606 11:51:43.639168       7 webrtc.go:277] Add Ice Candidate: candidate:15 1 TCP 2105524479 192.168.1.121 9 typ host tcptype active
cloud-game_1  | I0606 11:51:43.639267       7 webrtc.go:277] Add Ice Candidate: candidate:16 1 TCP 2105393407 172.17.0.1 9 typ host tcptype active
cloud-game_1  | I0606 11:51:43.639414       7 webrtc.go:277] Add Ice Candidate: candidate:5 1 UDP 2122121471 172.17.0.1 44393 typ host
cloud-game_1  | I0606 11:51:43.639678       6 browser.go:35] Browser 4c5de699-5309-40f9-b52b-63e801bd3a23] Received IceCandidate from browser -> relay to worker
cloud-game_1  | I0606 11:51:43.639809       7 coordinator.go:127] Received remote Ice Candidate from browser
cloud-game_1  | I0606 11:51:43.639845       7 webrtc.go:269] Decoded Ice: candidate:17 1 TCP 2105458943 172.18.0.1 9 typ host tcptype active
cloud-game_1  | I0606 11:51:43.639911       7 webrtc.go:277] Add Ice Candidate: candidate:17 1 TCP 2105458943 172.18.0.1 9 typ host tcptype active
cloud-game_1  | I0606 11:51:43.791758       6 browser.go:35] Browser 4c5de699-5309-40f9-b52b-63e801bd3a23] Received IceCandidate from browser -> relay to worker
cloud-game_1  | I0606 11:51:43.792261       7 coordinator.go:127] Received remote Ice Candidate from browser
cloud-game_1  | I0606 11:51:43.792334       7 webrtc.go:269] Decoded Ice: candidate:1 1 UDP 1686052863 95.30.109.77 39001 typ srflx raddr 192.168.1.121 rport 39001
cloud-game_1  | I0606 11:51:43.792492       7 webrtc.go:277] Add Ice Candidate: candidate:1 1 UDP 1686052863 95.30.109.77 39001 typ srflx raddr 192.168.1.121 rport 39001
cloud-game_1  | I0606 11:51:43.794222       7 webrtc.go:170] ICE Connection State has changed: connected
cloud-game_1  | I0606 11:51:43.794285       7 webrtc.go:174] ConnectionStateConnected
cloud-game_1  | I0606 11:51:43.794305       7 webrtc.go:309] Start streaming
cloud-game_1  | I0606 11:51:43.801457       6 browser.go:35] Browser 4c5de699-5309-40f9-b52b-63e801bd3a23] Received IceCandidate from browser -> relay to worker
cloud-game_1  | I0606 11:51:43.801624       7 coordinator.go:127] Received remote Ice Candidate from browser
cloud-game_1  | I0606 11:51:43.801657       7 webrtc.go:269] Decoded Ice: candidate:6 1 UDP 1685921791 95.30.109.77 44393 typ srflx raddr 172.17.0.1 rport 44393
cloud-game_1  | I0606 11:51:43.801722       7 webrtc.go:277] Add Ice Candidate: candidate:6 1 UDP 1685921791 95.30.109.77 44393 typ srflx raddr 172.17.0.1 rport 44393
cloud-game_1  | I0606 11:51:43.821208       7 webrtc.go:154] Data channel 'game-input'-'824635294554' open.
cloud-game_1  | I0606 11:51:43.829091       6 browser.go:35] Browser 4c5de699-5309-40f9-b52b-63e801bd3a23] Received IceCandidate from browser -> relay to worker
cloud-game_1  | I0606 11:51:43.829281       7 coordinator.go:127] Received remote Ice Candidate from browser
cloud-game_1  | I0606 11:51:43.829326       7 webrtc.go:269] Decoded Ice: candidate:11 1 UDP 1685987327 95.30.109.77 55024 typ srflx raddr 172.18.0.1 rport 55024
cloud-game_1  | I0606 11:51:43.829426       7 webrtc.go:277] Add Ice Candidate: candidate:11 1 UDP 1685987327 95.30.109.77 55024 typ srflx raddr 172.18.0.1 rport 55024
cloud-game_1  | I0606 11:51:47.202175       6 browser.go:35] Browser 4c5de699-5309-40f9-b52b-63e801bd3a23] Received start request from a browser -> relay to worker
cloud-game_1  | I0606 11:51:47.202445       7 coordinator.go:150] Received a start request from coordinator
cloud-game_1  | I0606 11:51:47.202550       7 coordinator.go:301] Starting game Sushi The Cat
cloud-game_1  | I0606 11:51:47.202599       7 coordinator.go:307] Got Room from local  <nil>  ID:  
cloud-game_1  | I0606 11:51:47.202629       7 room.go:78] Init new room:  45a8e84a3749e87c___Sushi The Cat Sushi The Cat {Sushi The Cat assets/games/Sushi The Cat.gba gba}
cloud-game_1  | I0606 11:51:47.202662       7 room.go:212] Updated player Index to:  0
cloud-game_1  | I0606 11:51:47.202690       7 coordinator.go:325] Is PC in room false
cloud-game_1  | I0606 11:51:47.202705       7 handlers.go:126] Detach peerconnection
cloud-game_1  | I0606 11:51:47.202894       7 room.go:223] Start WebRTC session
cloud-game_1  | I0606 11:51:47.202986       7 room.go:99] Check  /root/.cr/save/45a8e84a3749e87c___Sushi The Cat.dat  on online storage :  false
cloud-game_1  | I0606 11:51:47.203002       7 room.go:360] Check if game is on cloud storage
cloud-game_1  | I0606 11:51:47.203018       6 browser.go:35] Browser 4c5de699-5309-40f9-b52b-63e801bd3a23] Received room response from browser: 45a8e84a3749e87c___Sushi The Cat
cloud-game_1  | I0606 11:51:47.203133       7 room.go:105] Room 45a8e84a3749e87c___Sushi The Cat started. GamePath: assets/games/Sushi The Cat.gba, GameName: Sushi The Cat
cloud-game_1  | I0606 11:51:47.203158       6 worker.go:49] Coordinator: Received registerRoom room 45a8e84a3749e87c___Sushi The Cat from worker 6f9e1b72-d243-48ac-8cc0-b34f90498192
cloud-game_1  | I0606 11:51:47.203202       6 worker.go:51] Coordinator: Current room list is: map[45a8e84a3749e87c___Sushi The Cat:6f9e1b72-d243-48ac-8cc0-b34f90498192]
cloud-game_1  | Video pixel: {0 0 0 2 0 2 0} 2 0 1 2Libretro API version: 1
cloud-game_1  | ROM size: 540296
cloud-game_1  |   library_name: mGBA
cloud-game_1  |   library_version: 0.7.0 525d653
cloud-game_1  |   valid_extensions: gba|gb|gbc
cloud-game_1  |   need_fullpath: false
cloud-game_1  |   block_extract: false
cloud-game_1  | -----------------------------------
cloud-game_1  | --- System audio and video info ---
cloud-game_1  | -----------------------------------
cloud-game_1  |   Aspect ratio:  1.5
cloud-game_1  |   Base width:  240
cloud-game_1  |   Base height:  160
cloud-game_1  |   Max width:  240
cloud-game_1  |   Max height:  160
cloud-game_1  |   Sample rate:  32768
cloud-game_1  |   FPS:  59.727500915527344
cloud-game_1  | -----------------------------------
cloud-game_1  | listening audio channel true
cloud-game_1  | I0606 11:51:47.213735       7 room.go:127] Viewport custom size is disabled, base size will be used instead 240x160
cloud-game_1  | I0606 11:51:47.213767       7 room.go:134] meta:  {assets/emulator/libretro/cores/mgba_libretro 240 160 32768 59 240 160 1.5 {<nil> false}}
cloud-game_1  | I0606 11:51:47.213981       7 media.go:138] Video Encoder:  VP8
cloud-game_1  | I0606 11:51:47.214089       7 media.go:73] Enter fan audio
```

**Comment by sergystepanov** [comment]
@giongto35, can you try this Dockerfile with compose if you will (`docker-compose up` and if you need rebuild -- `docker-compose build` then `up`). To check if this one is OK.
Not sure what to do with `assets`. At the moment they are ignored and suppose to be mounted manually. For Docker image releases should we pack cores and just ignore all games?
As you may know, Docker volume mounts will override any existing in the container folder, if pack some games into the image.

**Comment by giongto35** [comment]
I tried running docker-compose up but "/usr/local/share/cloud-game/worker: No such file or directory". However, docker build and docker run succeed.
Yes, Docker image should be with just cores and ignore games.

**Comment by giongto35** [comment]
@sergystepanov  Oh it's the image name. I can run it now . Thank you for the good work : D .

**Comment by giongto35** [comment]
Still need the cores to be attached to get it fully work right?

**Comment by sergystepanov** [comment]
Try it right now. build and then up.

**Comment by sergystepanov** [comment]
I made it to pack all the cores into the image. +160Mb for an image.

image

**Comment by sergystepanov** [comment]
Tested on Ubuntu and it's working. Don't forget now to clear old containers that use cloud-game-local image.
Gonna read gh docker repo docs for image release. (%

File: docker-compose.yml [hunk 1]
```diff
@@ -3,7 +3,7 @@ services:
 
   cloud-game:
     build: .
-    image: "cloud-game"
+    image: "cloud-game-local"
```
[the remark this answers is not in the document]
Comment (sergystepanov) [reply docker-compose.yml]: ???

File: .dockerignore [hunk 2]
```diff
@@ -1,6 +1,16 @@
-*/assets/games/*
-.git
+.git/
 .gitignore
+.github/
+.idea/
+
+.env
+Dockerfile
```
[the remark this answers is not in the document]
Comment (sergystepanov) [reply .dockerignore:7]: dockerignore filters files for initial build context. We don't need them inside the build image. Not sure what you meant, but you can update them.

[the remark this answers is not in the document]
Comment (sergystepanov) [reply .dockerignore:7]: Also you can even add .dockerignore into itself.

[the remark this answers is not in the document]
Comment (giongto35) [reply .dockerignore:7]: ah sorry, mistake with gitignore : D.

**Pull request merged**